src/txn/txn.c

```diff
@@ -1291,23 +1333,42 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
             hs_recno_key.size = WT_PTRDIFF(p, hs_recno_key_buf);
             hs_cursor->set_key(hs_cursor, 4, btree->id, &hs_recno_key, WT_TS_MAX, UINT64_MAX);
         }
+        /*
+         * Locate the previous update from the history store. We know there may be content in the
+         * history store if the prepared update is written to the disk image or first committed
+         * update older than the prepared update is marked as WT_UPDATE_HS. The second case is rare
+         * but can happen if the previous eviction that writes the prepared update to the disk image
+         * fails after reconciliation.
+         *
+         * We need to locate the history store update before we resolve the prepared updates because
+         * if we abort the prepared updates first, the history store search may race with other
+         * sessions modifying the same key and checkpoint moving the new updates to the history
+         * store.
+         */
         WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_before(session, hs_cursor), true);
 
         /* We should only get not found if the prepared update is on disk. */
-        WT_ASSERT(session, ret != WT_NOTFOUND || prepare_on_disk);
-        if (ret == WT_NOTFOUND && !commit) {
+        WT_ASSERT(session, ret != WT_NOTFOUND || resolve_case == RESOLVE_PREPARE_ON_DISK);
+        if (ret == 0) {
+            has_hs_record = true;
```

Comment:
This fixes the issue in the previous version that we will fix the history store record even there is none history store record.